examples/cpp/grpc/async-server/async-server.cpp

```diff
@@ -0,0 +1,129 @@
+#include <bond/ext/grpc/unary_call.h>
+#include <bond/ext/grpc/wait_callback.h>
+
+#include <chrono>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <thread>
+
+using grpc::Channel;
+using grpc::ServerBuilder;
+using grpc::Status;
+
+using namespace helloworld;
+
+class GreeterServiceImpl final : public Greeter::Service
+{
+public:
+    // For this example, we will use the same thread pool to perform
+    // asynchronous processing of requests as is used in the rest of the
+    // program.
+    explicit GreeterServiceImpl(std::shared_ptr<bond::ext::gRPC::thread_pool> tp)
+        : _tp(std::move(tp))
+    { }
+
+private:
+    using CallType = bond::ext::gRPC::unary_call<bond::bonded<HelloRequest>, HelloReply>;
+    using SharedCallType = bond::ext::gRPC::shared_unary_call<bond::bonded<HelloRequest>, HelloReply>;
+
+    std::shared_ptr<bond::ext::gRPC::thread_pool> _tp;
+
+    class PerRequestState
+    {
+    public:
+        explicit PerRequestState(CallType call)
+        : _call(std::move(call).share())
+        { }
+
+        // In this example, the PerRequestState doubles as the thread pool
+        // callback. This need not be the case.
+        void operator()()
+        {
+            // Simulate doing some other work...
+            std::this_thread::sleep_for(std::chrono::milliseconds(750));
+
+            HelloRequest request = _call.request().Deserialize();
+
+            HelloReply reply;
+            reply.message = "hello " + request.name;
+
+            _call.Finish(reply);
+        }
+
+    private:
+        // The thread pool implementation that we're using requires its
+        // Callbacks to be copyable, so we switch to using
+        // shared_unary_call, which is copyable.
+        SharedCallType _call;
+
+        // Other state could be added as needed.
+    };
+
+    void SayHello(CallType call) override
+    {
+        PerRequestState perRequestState { std::move(call) };
+        _tp->schedule(perRequestState);
+    }
+};
+
+int main()
+{
+    auto ioManager = std::make_shared<bond::ext::gRPC::io_manager>();
+    auto threadPool = std::make_shared<bond::ext::gRPC::thread_pool>();
+
+    GreeterServiceImpl service(threadPool);
+
+    bond::ext::gRPC::server_builder builder;
+    builder.SetThreadPool(threadPool);
+    const std::string server_address("127.0.0.1:50051");
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+    builder.RegisterService(&service);
+    std::unique_ptr<bond::ext::gRPC::server> server(builder.BuildAndStart());
+
+    Greeter::Client greeter(
+        grpc::CreateChannel(server_address, grpc::InsecureChannelCredentials()),
+        ioManager,
+        threadPool);
+
+    const std::string user("world");
+
+    HelloRequest request;
+    request.name = user;
+
+    bond::ext::gRPC::wait_callback<HelloReply> cb;
+    greeter.AsyncSayHello(request, cb);
+
+    bool waitResult = cb.wait_for(std::chrono::seconds(10));
```

Comment:
Too many descriptive variables sometimes has the opposite effect on readability in my opinion. The `if (!cb.wait_for(...))` seems cleaner.